packages/opencensus-web-instrumentation-zone/test/test-interaction-tracker.ts

```diff
@@ -0,0 +1,408 @@
+      expect(incrementTaskCountSpy.calls.count()).toBe(2);
+      expect(decrementTaskCountSpy.calls.count()).toBe(2);
+      expect(rootSpan.name).toBe('Navigation /test_navigation');
+      expect(rootSpan.ended).toBeTruthy();
+      expect(rootSpan.duration).toBeLessThanOrEqual(10);
+      done();
+    });
+  });
+
+  it('should handle HTTP requets', done => {
+    const onclick = () => {
+      expect(incrementTaskCountSpy.calls.count()).toBe(1);
+      expect(decrementTaskCountSpy.calls.count()).toBe(0);
+      doHTTPRequest();
+    };
+    fakeInteraction(onclick);
+
+    onEndSpanSpy.and.callFake((rootSpan: Span) => {
+      expect(incrementTaskCountSpy.calls.count()).toBe(2);
+      expect(decrementTaskCountSpy.calls.count()).toBe(2);
+      expect(rootSpan.name).toBe('test interaction');
+      expect(rootSpan.ended).toBeTruthy();
+      expect(rootSpan.duration).toBeGreaterThanOrEqual(60);
+      expect(rootSpan.duration).toBeLessThanOrEqual(80);
+      done();
+    });
+  });
+
+  it('should handle cascading tasks', done => {
+    const onclick = () => {
+      expect(incrementTaskCountSpy.calls.count()).toBe(1);
+      expect(decrementTaskCountSpy.calls.count()).toBe(0);
+      const promise = getPromise();
+      promise.then(() => {
+        expect(incrementTaskCountSpy.calls.count()).toBe(2);
+        expect(decrementTaskCountSpy.calls.count()).toBe(1);
+        setTimeout(() => {
+          expect(incrementTaskCountSpy.calls.count()).toBe(3);
+          expect(decrementTaskCountSpy.calls.count()).toBe(2);
+          doHTTPRequest();
+        }, 1);
+      });
+    };
+    fakeInteraction(onclick);
+
+    onEndSpanSpy.and.callFake((rootSpan: Span) => {
+      expect(incrementTaskCountSpy.calls.count()).toBe(4);
+      expect(decrementTaskCountSpy.calls.count()).toBe(4);
+      expect(rootSpan.name).toBe('test interaction');
+      expect(rootSpan.ended).toBeTruthy();
+      // The HTTP request takes 60 ms to send it
+      expect(rootSpan.duration).toBeGreaterThanOrEqual(60);
+      expect(rootSpan.duration).toBeLessThanOrEqual(80);
+      done();
+    });
+  });
+
+  function fakeInteraction(callback: Function, elem?: HTMLElement) {
+    let element: HTMLElement;
+    if (elem) element = elem;
+    else element = createButton('test interaction');
+    element.onclick = () => callback();
+    element.click();
+  }
+
+  function createButton(dataOcwebId: string, disabled?: boolean): HTMLElement {
+    const button = document.createElement('button');
+    button.setAttribute('data-ocweb-id', dataOcwebId);
+    button.setAttribute('id', 'test_element');
+    if (disabled) {
+      button.setAttribute('disabled', 'disabled');
+    }
+    return button;
+  }
+
+  const noop = () => {};
+
+  function getPromise() {
+    return new Promise(resolve => {
+      resolve();
+    });
+  }
+
+  function doHTTPRequest() {
+    const xhr = new XMLHttpRequest();
+    spyOn(xhr, 'send').and.callFake(() => {
+      setTimeout(noop, 60);
```

Comment:
What if this `60` was a constant like XHR_TIME? And what if we also had a `SET_TIMEOUT_TIME`? Then in the tests you could have something that adds up the times so something like this:

```
const interactionTime = XHR_TIME + SET_TIMEOUT_TIME;
expect(rootSpan.duration).toBeGreaterThanOrEqual(interactionTime);
expect(rootSpan.duration).toBeLessThanOrEqual(interactionTime + BUFFER_TIME);
```